0000-zmix/README.md

```diff
@@ -0,0 +1,395 @@
+- Signatures
+    - Boneh Boyen Shacham
+    - Pointcheval Saunders
+- Pedersen commitments
+- Bulletproof intervals
+- Bulletproof set membership inclusive and exclusive
+- zk-SNARK set memberships
+- Verifiable encryption
+- Linkable indistinguisable tags (see ia.cr/2011/658) to realize scope pseudonyms
+
+## Structures
+
+### Statements
+
+```rust
+pub enum Statement {
+    /// Boneh Boyen Shacham Signature
+    SignatureBBS {
+        pk: Vec<u8>,
+        messages: Vec<HiddenOrRevealedValue>,
+    },
+    /// Pointcheval Sanders Signature
+    SignaturePS {
+        pk: Vec<u8>,
+        messages: Vec<HiddenOrRevealedValue>,
+    },
+    PedersenCommitment {
+        message_index: usize,
+        commitment: Vec<u8>,
+        params: PedersenCommitmentParams,
+    },
+    IntervalBulletproof {
+        message_index: usize,
+        min: Vec<u8>,
+        max: Vec<u8>,
+        params: Vec<u8>,
+    },
+    /// Camenisch Shoup Encryption
+    EncryptionCS {
+        message_index: usize,
+        pk: Vec<u8>,
+        ciphertext: Vec<u8>,
+    },
+    /// As defined by Bernhard et al., Anonymous attestation with user-controlled linkability (ia.cr/2011/658)
+    LinkableIndistinguishableTagBLS {
+        message_index: usize,
+        tag: Vec<u8>,
+        params: Vec<u8>,
+    },
+}
+
+pub enum HiddenOrRevealedValue {
+    HiddenValueIndex(usize),
+    RevealedValue(Vec<u8>),
+}
+
+pub struct PedersenCommitmentParams(pub Vec<u8>);
+```
+
+### Proof Specification
+
+The proof spec follows this data model
+```rust
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ProofSpec {
+    #[serde(rename = "messageCount")]
+    pub message_count: usize,
+    pub statements: Vec<Statement>,
+    pub params: ProofSpecParams,
+}
+
+pub enum ProofSpecParams {
+    BN254,
+    BLS12_381,
+    Ed25519,
+}
+```
+
+For example, a proof spec could be represented in JSON as
+```json
+{
+  "messageCount": 2,
+  "statements": [
+    {
+      "type": "SignatureBBS",
+      "data": {
+        "pk": [
+          112,
+          117,
+          98,
+          108,
+          105,
+          99,
+          107,
+          101,
+          121
+        ],
+        "messages": [
+          {
+            "hiddenValueIndex": 0
+          },
+          {
+            "revealedValue": [
+              118,
+              97,
+              108,
+              117,
+              101
+            ]
+          }
+        ]
+      }
+    }
+  ],
+  "params": "BLS12_381"
+}
+
+```
+
+### Witness
+The witness follows this data model
+
+```rust
+/// A witness w is valid w.r.t. a proof specification s if
+/// * w contains one value for each of the (hidden) messages in s, so w.messages.len() == s.message_count
+/// * w contains one statement witness for each statement in s, so w.statement_witnesses.len() == s.statements.len()
+/// * TODO: add further requirements
+pub struct Witness {
+    messages: Vec<Vec<u8>>,
+    statement_witnesses: Vec<StatementWitness>,
```

Comment:
What is the purpose of `statement_witness` (and the use of a struct) vs just having a witness list?

Reply:
Not sure I completely understand the question but let me try to clarify: We get a vector of witness information such that each piece of the witness information corresponds to one of the proof statements. It's a struct such that we can make it structured data, for example, a BBS signature BBS proof expects a BBS signature as a witness, so we use a StatementWitness struct rather than just bytes.

Reply:
Will remove `statement_witness` and just leave `messages`